overlord/servicestate/quota_handlers.go

```diff
@@ -37,6 +40,302 @@ import (
 	"github.com/snapcore/snapd/wrappers"
 )
 
+// QuotaControlAction is the serialized representation of a quota group
+// modification that lives in a task.
+type QuotaControlAction struct {
+	// QuotaName is the name of the quota group being controlled.
+	QuotaName string `json:"quota-name"`
+
+	// Action is the action being taken on the quota group. It can be either
+	// "create", "update", or "remove".
+	Action string `json:"action"`
+
+	// AddSnaps is the set of snaps to add to the quota group, valid for either
+	// the "update" or the "create" actions.
+	AddSnaps []string `json:"snaps"`
+
+	// MemoryLimit is the memory limit for the quota group being controlled,
+	// either the initial limit the group is created with for the "create"
+	// action, or if non-zero for the "update" the memory limit, then the new
+	// value to be set.
+	MemoryLimit quantity.Size
+
+	// ParentName is the name of the parent for the quota group if it is being
+	// created. Eventually this could be used with the "update" action to
+	// support moving quota groups from one parent to another, but that is
+	// currently not supported.
+	ParentName string
+}
+
+func (m *ServiceManager) doQuotaControl(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	defer perfTimings.Save(st)
+
+	meter := snapstate.NewTaskProgressAdapterUnlocked(t)
+
+	qcs := []QuotaControlAction{}
+	err := t.Get("quota-control-actions", &qcs)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot get quota-control-action: %v", err)
+	}
+
+	// TODO: support more than one action
+	switch {
+	case len(qcs) > 1:
+		return fmt.Errorf("multiple quota group actions not supported yet")
+	case len(qcs) == 0:
+		return fmt.Errorf("internal error: no quota group actions for quota-control task")
+	}
+
+	qc := qcs[0]
+
+	allGrps, err := AllQuotas(st)
+	if err != nil {
+		return err
+	}
+
+	args := quotaControlArgs{
+		St:          st,
+		AllGrps:     allGrps,
+		Action:      &qc,
+		Meter:       meter,
+		PerfTimings: perfTimings,
+	}
+
+	switch qc.Action {
+	case "create":
+		err = doCreateQuota(args)
+	case "remove":
+		err = doRemoveQuota(args)
+	case "update":
+		err = doUpdateQuota(args)
+	default:
+		err = fmt.Errorf("unknown action %q requested", qc.Action)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	t.SetStatus(state.DoneStatus)
+
+	return nil
+}
+
+func doCreateQuota(args quotaControlArgs) error {
```

Comment:
We might have one or two exceptions (I couldn't remember where), but I think we generally use doSomething names for task handlers, so maybe just createQuota, removeQuota etc?

Reply:
are these functions not handlers? I assumed that the implementation of a task was the thing that was called `doSomething` and the `something()` is the bit that returned the task, so we will have `servicestate.CreateQuota()` which returns a task, and the handler for the task is `doCreateQuota`

Reply:
oh sorry I see what you mean now, the task is `quota-control` and I have doQuotaControl as the handlers, these are really handler helpers not proper handlers. I personally dislike having two functions which differ only by their capitalization and lack thereof, so I will rename it something other than just `createQuota` etc but thanks for pointing this out

Reply:
I renamed this

Reply:
well I forgot to push it yesterday, but I will push it today with the other updates

Reply:
this is pushed now